lib/AST/Decl.cpp

```diff
@@ -636,9 +636,18 @@ const ExternalSourceLocs *Decl::getSerializedLocs() const {
     return &NullLocs;
   }
 
+  CharSourceRange BufferRange = SM.getRangeForBuffer(BufferID);
   auto ResolveLoc = [&](const ExternalSourceLocs::RawLoc &Raw) -> SourceLoc {
+    // If the underlying source has been updated and the swiftsourceinfo hasn't,
+    // make sure we don't produce invalid source locations. Ideally would check
+    // the file hasn't been modified.
+    if (Raw.Offset > BufferRange.getByteLength())
```

Comment:
Should this be a `>=`? I believe if `Raw.Offset` was equal to `BufferRange.getByteLength()`, that would cause us to form a `SourceLoc` to the trailing `\0` in the buffer, which doesn't seem like it should be valid.

Reply:
Ah okay!